hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -433,84 +434,96 @@ def freeze(self):
-            buf = hbytes(buf)
-            self.__write(buf)
-            result = int_from_bytes(buf)
-
-        assert bit_length(result) <= n
-        return result
-
-    def write(self, string):
-        string = hbytes(string)
-        self.__assert_not_frozen("write")
-        self.__check_capacity(len(string))
-        original = self.index
-        self.__write(string, forced=True)
-        self.forced_indices.update(hrange(original, self.index))
-        return string
-
-    def __check_capacity(self, n):
-        if self.index + n > self.max_length:
-            self.overdraw = self.index + n - self.max_length
-            self.status = Status.OVERRUN
-            self.freeze()
-            raise StopTest(self.testcounter)
+        buf = hbytes(buf)
+        result = int_from_bytes(buf)
 
-    def __write(self, result, forced=False):
         self.start_example(DRAW_BYTES_LABEL)
         initial = self.index
-        n = len(result)
 
         block = Block(
             start=initial,
-            end=initial + n,
+            end=initial + n_bytes,
             index=len(self.blocks),
-            forced=forced,
-            all_zero=not any(result),
+            forced=forced is not None,
+            all_zero=result == 0,
         )
 
-        self.block_starts.setdefault(n, []).append(block.start)
+        self.block_starts.setdefault(n_bytes, []).append(block.start)
+
+        if block.forced:
+            self.forced_indices.update(hrange(block.start, block.end))
+        self.block_starts.setdefault(n_bytes, []).append(block.start)
         self.blocks.append(block)
         assert self.blocks[block.index] is block
-        assert len(result) == n
         assert self.index == initial
-        self.buffer.extend(result)
+        self.buffer.extend(buf)
         self.index = len(self.buffer)
         self.stop_example()
 
+        assert bit_length(result) <= n
+        return result
+
     def draw_bytes(self, n):
-        self.__assert_not_frozen("draw_bytes")
-        if n == 0:
-            return hbytes(b"")
-        self.__check_capacity(n)
-        result = self._draw_bytes(self, n)
-        assert len(result) == n
-        self.__write(result)
-        return hbytes(result)
+        """Draw n bytes from the underlying source."""
+        return int_to_bytes(self.draw_bits(8 * n), n)
 
-    def mark_interesting(self, interesting_origin=None):
-        self.__assert_not_frozen("mark_interesting")
+    def write(self, string):
+        """Write ``string`` to the output buffer."""
+        self.__assert_not_frozen("write")
+        string = hbytes(string)
+        if not string:
+            return
+        self.draw_bits(len(string) * 8, forced=int_from_bytes(string))
+        return self.buffer[-len(string) :]
+
+    def __check_capacity(self, n):
+        if self.index + n > self.max_length:
+            self.overdraw = self.index + n - self.max_length
+            self.status = Status.OVERRUN
```

Comment:
Could this use `self.comclude_test`?

Reply:
It could! It could also stop assigning to a field that we never use anywhere (I wonder what that was for, but I'm not curious enough to dive into history and find out).

Reply:
It originally dates back to https://github.com/HypothesisWorks/hypothesis/commit/095575bb88e025aa568aafbae624541f993fde72 in early 2016, and seems to have been a crude heuristic for deciding that a new overrun result is “better” than a previous one because it overran by a smaller amount.

Its usage then disappeared in #880, specifically in https://github.com/HypothesisWorks/hypothesis/commit/47b0dc725dc207913f136db1fa9aa59569d31cf3, in favour of a simple status comparison with no additional ranking of overruns.

It does not seem to have been terribly important.